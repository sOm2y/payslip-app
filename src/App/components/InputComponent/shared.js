//@flow

import React from 'react';

import { FormHelperText } from 'material-ui/Form';

function renderFailText(failReason: string) {
  if (failReason === null || typeof failReason === 'undefined') {
    throw new Error('fail reason cannot be empty, please check error message render method in App/InputComponent/shared.js');
  }

  return <FormHelperText>{failReason}</FormHelperText>;
}

function doMapState(state: InputComponentStates): PayslipInput {
  if (
    state === null ||
    typeof state === 'undefined' ||
    state.annualSalary === null ||
    typeof state.annualSalary === 'undefined'
  ) {
    throw new Error('empty properties found in `InputComponentState`, please check state input, got: ', JSON.stringify(state));
  }

  return {
    firstname: state.firstname,
    lastname: state.lastname,
    annualSalary: parseInt(state.annualSalary.replace(',', ''), 0),
    superRate: parseInt(state.superRate, 0) / 100,
    paymentStartDate: state.paymentStartDate
  };
}

export { renderFailText, doMapState };
