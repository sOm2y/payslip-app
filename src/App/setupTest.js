//@flow

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// enable using of expect(wrapper).toContainReact(<SomeComponent/>)
import 'jest-enzyme';

configure({ adapter: new Adapter() });
