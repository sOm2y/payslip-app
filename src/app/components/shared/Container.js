//@flow

import React from 'react';
import type { Node } from 'react';

import Paper from 'material-ui/Paper';
import { withStyles } from 'material-ui/styles';

const styles = {
  container: {
    padding: 32,
    maxWidth: 360
  }
};

class Container extends React.PureComponent<{ children: Node, classes: { container: string } }> {
  render() {
    const { classes, children, ...others } = this.props;

    return (
      <Paper {...others} className={classes.container}>
        {children}
      </Paper>
    );
  }
}

export default withStyles(styles)(Container);
