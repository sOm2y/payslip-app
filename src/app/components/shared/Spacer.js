//@flow

import React from 'react';

class Spacer extends React.PureComponent<{ height?: number, width?: number }> {
  render() {
    const { height, width } = this.props;

    return (
      <div
        style={{
          height,
          width
        }}
      />
    );
  }
}

export default Spacer;
