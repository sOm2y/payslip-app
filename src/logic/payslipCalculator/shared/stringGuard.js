//@flow

/**
 * shoud return false when input is empty/null/undefined
 */

function rejectEmptyNullUndefined(input: string) {
  return rejectEmpty(input);
}

function rejectUndefined(input: string) {
  return typeof input === 'undefined' ? true : false;
}

function rejectNull(input: string) {
  return input === null ? true : false;
}

function rejectEmpty(input: string) {
  return rejectUndefined(input) || rejectNull(input) || typeof input !== 'string' || input.length === 0;
}

export { rejectEmptyNullUndefined };
