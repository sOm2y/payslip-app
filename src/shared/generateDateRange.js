//@flow

import format from 'date-fns/format';
import addMonths from 'date-fns/add_months';
import startOfMonth from 'date-fns/start_of_month';
import subtractDays from 'date-fns/sub_days';
import getYear from 'date-fns/get_year';
import getMonth from 'date-fns/get_month';

function formatDate(date: Date) {
  return format(date, 'Do MMMM');
}

function from0To11() {
  return Array.from(Array(12).keys());
}

function firstDayOfCurrentMonth() {
  const today = new Date();

  return startOfMonth(today);
}

/**
 * put things together
 */
function generateDateRange(
  from0To11: () => number[],
  getFirstDayOfThisMonth: () => Date,
  addMonths: (Date, number) => Date,
  subtractDays: (Date, number) => Date,
  getYear: Date => number,
  getDateString: Date => string
) {
  const firstDayOfThisMonth = getFirstDayOfThisMonth();

  return (): MonthAndRange[] => {
    return from0To11().map((increment) => {
      const firstDay = addMonths(firstDayOfThisMonth, increment);
      const lastDay = subtractDays(addMonths(firstDay, 1), 1);
      const thisYear = getYear(firstDay);

      return {
        month: getMonth(firstDay).toString(),
        range: `${getDateString(firstDay)} - ${getDateString(lastDay)}, ${thisYear}`
      };
    });
  };
}

export default generateDateRange(from0To11, firstDayOfCurrentMonth, addMonths, subtractDays, getYear, formatDate);
