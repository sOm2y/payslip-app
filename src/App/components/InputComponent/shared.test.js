//@flow

import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import '../../setupTest'; // enzyme setup
import { renderFailText, doMapState } from './shared';

describe('shared', () => {
  describe('renderFailText', () => {
    it('should render fail message', () => {
      const failMessage = 'foo';
      const element = shallow(renderFailText(failMessage));
      expect(element.html()).toContain(failMessage);
    });

    it('should throw error when input is null/undefined', () => {
      expect(() => renderFailText(undefined)).toThrow();
      expect(() => renderFailText(null)).toThrow();
    });

    it('should match snapshot', () => {
      const tree = renderer.create(renderFailText('fail')).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('doMapState', () => {
    it('should map state data', () => {
      const state = {
        firstname: 'foo',
        lastname: 'bar',
        annualSalary: '30000',
        superRate: '10',
        paymentStartDate: 'wola'
      };

      const result = doMapState(state);

      expect(result.firstname).toEqual(state.firstname);
      expect(result.lastname).toEqual(state.lastname);
      expect(result.paymentStartDate).toEqual(state.paymentStartDate);
      expect(result.annualSalary).toEqual(30000);
      expect(result.superRate).toEqual(0.1);
    });

    it('should throw error when input is null/undefined', () => {
      expect(() => doMapState(null)).toThrow();
      expect(() => doMapState(undefined)).toThrow();
    });

    it('should throw error when state.annualSalary is null/undefined', () => {
      const state = {
        firstname: 'foo',
        lastname: 'bar',
        annualSalary: '30000',
        superRate: '10',
        paymentStartDate: 'wola'
      };

      expect(() => doMapState({ ...state, annualSalary: null })).toThrow();
      expect(() => doMapState({ ...state, annualSalary: undefined })).toThrow();
    });
  });
});
