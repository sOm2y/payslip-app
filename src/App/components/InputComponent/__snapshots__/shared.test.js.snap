// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared renderFailText should match snapshot 1`] = `
<p
  className="MuiFormHelperText-root-1"
>
  fail
</p>
`;
